src/util/address.rs

```diff
@@ -354,6 +746,83 @@ impl From<WitnessVersion> for opcodes::All {
     }
 }
 
+/// Structure providing information and validity guarantees for witness programs as parts of
+/// bitcoin addresses and `scriptPubkey` scripts
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct WitnessProgram {
+    /// The witness program version
+    version: WitnessVersion,
+    /// The witness program
+    program: Vec<u8>,
+}
+
+impl WitnessProgram {
+    /// Extracts witness program from a given script (assumed to be a `scriptPubkey`). Errors if
+    /// the script is not a valid non-legacy SegWit scriptPubkey
+    pub fn from_script(script: &script::Script) -> Result<WitnessProgram, WitnessError> {
+        if !script.is_witness_program() {
+            return Err(WitnessError::NotSegwitScriptPubkey)
+        }
+        let version = WitnessVersion::from_opcode(opcodes::All::from(script[0]))?;
+        let program = script[2..].to_vec();
```

Comment:
is it guaranteed script[1] is a OP_PUSHDATA using 1 byte?

Reply:
The data type is removed, so the question is solved